End TritonBundle on non-cache write codepaths

oulgen opened the pull request:
Summary:
When we bypass cache write on inductor, we were also forgetting to reset the bundle, this moves resetting the bundle into post_compile step so it gets uniformly reset.

This diff also turns on the cache for internal so that we can do a code rollout.

Test Plan: updated tests

Differential Revision: D65457224




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -105,6 +109,14 @@ def begin_compile(cls) -> None:
+        Finalizes the TritonBundler. If collect is not yet called, it
+        discards the current bundle.
+        """
+        cls._entries = None
```
This is not blocking this PR since you're in a rush, but this is something where it's suuuuper useful to have debug() log saying that end_compile was called.

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -171,8 +183,9 @@ def collect(
-                cls._entries = None
+                cls.end_compile()
                 return result, TritonBundlerMetadata(kernel_names)
+            cls.end_compile()
```
This is pointless right? You JUST tested that entries is None.

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1217,6 +1218,10 @@ def post_compile(
         set_tracing_context_output_strides(example_inputs, compiled_graph)
 
+        # If we have not ended compile using TritonBundler.collect, lets
+        # end here.
```
@aorenste Doesn't this mean Triton autotune bundled cache also has this same problem?

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1217,6 +1218,10 @@ def post_compile(
 
+        # If we have not ended compile using TritonBundler.collect, lets
+        # end here.
+        TritonBundler.end_compile()
```
The current scheme is not exception safe. If compilation raises an exception you will fail to clear TritonBundler. If the exception is suppressed with `suppress_errors` you will then trigger an unrelated error on subsequent compilation.

Why do you not just hoist the `begin_compile` to codecache.py load. You can then very simply write:

```
with TritonBundler() as bundler:
  ... = codegen_and_compile()
result = bundler.collect()
... write rest of cache ...
```

All other begin/compile annotations can be removed. Very simple, yes yes.

ezyang reviewed (changes requested):
see comment

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -171,7 +185,7 @@ def collect(
                             )
                         )
-                cls._entries = None
+                cls.end_compile()
```
You don't need these anymore, although it is harmless

oulgen replied:
i guess i do not need collect to end the compile as well anymore

oulgen committed:
End TritonBundle on non-cache write codepaths (#139698)

Summary:

When we bypass cache write on inductor, we were also forgetting to reset the bundle, this moves resetting the bundle into post_compile step so it gets uniformly reset.

This diff also turns on the cache for internal so that we can do a code rollout.

Test Plan: updated tests

Reviewed By: ezyang

Differential Revision: D65457224
